In [ ]:
from google.colab import files
uploaded = files.upload()

Saving Sellers_New.csv to Sellers_New.csv
Saving GeoLocation_New.csv to GeoLocation_New.csv


In [3]:
import pandas as pd
import numpy as np

In [105]:
import pandas as pd


GeoLocation_New = pd.read_csv("GeoLocation_New.csv")

In [49]:
import pandas as pd

Order_Payments_New = pd.read_csv("Order_Payments_New.csv")


In [84]:
import pandas as pd

Sellers_New = pd.read_csv("Sellers_New.csv")

In [2]:
import pandas as pd

Order_Reviews_New = pd.read_csv("Order_Reviews_New.csv")
Order_Payments_New = pd.read_csv("Order_Payments_New.csv")
Product_New = pd.read_csv("Product_New.csv")
Order_Item_New = pd.read_csv("Order_Item_New.csv")
Orders_New = pd.read_csv("Orders_New.csv")
Customers_New = pd.read_csv("Customers_New.csv")



FileNotFoundError: [Errno 2] No such file or directory: 'Order_Reviews_New.csv'

In [11]:
import os
os.listdir()

['.config',
 'Order_Reviews_New.csv',
 'GeoLocation_New.csv',
 'Customers_New.csv',
 'Orders_New.csv',
 'Order_Payments_New.csv',
 'Product_New.csv',
 'Sellers_New.csv',
 'Order_Item_New.csv',
 'sample_data']

In [ ]:
Order_Reviews_New.head()


,review_id,order_id,review_score,review_creation_date,review_answer_timestamp,response_time_hours
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,2018-01-18,2018-01-18 21:46:00,21.77
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,2018-03-10,2018-03-11 03:05:00,27.08
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,2018-02-17,2018-02-18 14:36:00,38.60
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,2017-04-21,2017-04-21 22:02:00,22.03
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,2018-03-01,2018-03-02 10:26:00,34.43


In [12]:
Order_Payments_New.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [ ]:
Sellers_New.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [13]:
GeoLocation_New.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [ ]:
Product_New.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [ ]:
Order_Item_New.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,9/19/2017 9:45,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,5/3/2017 11:05,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,1/18/2018 14:48,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,8/15/2018 10:10,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2/13/2017 13:57,199.90,18.14


In [ ]:
Orders_New.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,10/2/2017 10:56,10/2/2017 11:07,10/4/2017 19:55,10/10/2017 21:25,10/18/2017 0:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,7/24/2018 20:41,7/26/2018 3:24,7/26/2018 14:31,8/7/2018 15:27,8/13/2018 0:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,8/8/2018 8:38,8/8/2018 8:55,8/8/2018 13:50,8/17/2018 18:06,9/4/2018 0:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,11/18/2017 19:28,11/18/2017 19:45,11/22/2017 13:39,12/2/2017 0:28,12/15/2017 0:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2/13/2018 21:18,2/13/2018 22:20,2/14/2018 19:46,2/16/2018 18:17,2/26/2018 0:00


In [ ]:
Customers_New.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [ ]:
Order_Reviews_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [ ]:
Order_Payments_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [ ]:
GeoLocation_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [ ]:
Sellers_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [ ]:
Product_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [ ]:
Order_Item_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [ ]:
Product_New.head()

In [ ]:
Orders_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [ ]:
Customers_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [ ]:
Order_Reviews_New.isnull().sum()

,0
review_id,0
order_id,0
review_score,0
review_comment_title,87656
review_comment_message,58247
review_creation_date,0
review_answer_timestamp,0


In [ ]:
# i drop colum review_comment_title and review_comment_message because too much null values also don,'t require for analysis
Order_Reviews_New = Order_Reviews_New.drop(columns = ['review_comment_title','review_comment_message'])

In [ ]:
Order_Reviews_New.head()

,review_id,order_id,review_score,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,1/18/2018 0:00,1/18/2018 21:46
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,3/10/2018 0:00,3/11/2018 3:05
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,2/17/2018 0:00,2/18/2018 14:36
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,4/21/2017 0:00,4/21/2017 22:02
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,3/1/2018 0:00,3/2/2018 10:26


In [ ]:
Order_Reviews_New['review_creation_date'] = pd.to_datetime(Order_Reviews_New['review_creation_date'])
Order_Reviews_New['review_answer_timestamp'] = pd.to_datetime(Order_Reviews_New['review_answer_timestamp'])

In [ ]:
Order_Reviews_New.info()

<class 'pandas.core.frame.DataFrame'>
Index: 98410 entries, 0 to 99223
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                98410 non-null  object        
 1   order_id                 98410 non-null  object        
 2   review_score             98410 non-null  int64         
 3   review_creation_date     98410 non-null  datetime64[ns]
 4   review_answer_timestamp  98410 non-null  datetime64[ns]
 5   response_time_hours      98410 non-null  float64       
dtypes: datetime64[ns](2), float64(1), int64(1), object(2)
memory usage: 5.3+ MB


In [ ]:
Order_Reviews_New['response_time_hours'] = (Order_Reviews_New['review_answer_timestamp'] - Order_Reviews_New['review_creation_date']).dt.total_seconds() / 3600

In [ ]:
Order_Reviews_New.head()

,review_id,order_id,review_score,review_creation_date,review_answer_timestamp,response_time_hours
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,2018-01-18,2018-01-18 21:46:00,21.77
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,2018-03-10,2018-03-11 03:05:00,27.08
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,2018-02-17,2018-02-18 14:36:00,38.60
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,2017-04-21,2017-04-21 22:02:00,22.03
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,2018-03-01,2018-03-02 10:26:00,34.43


In [ ]:
Order_Reviews_New['response_time_hours'] = Order_Reviews_New['response_time_hours'].round(2)

In [ ]:
Order_Reviews_New.info()

In [ ]:
Order_Reviews_New['review_id'].str.len().unique()

array([32])

In [ ]:
Order_Reviews_New['order_id'].str.len().unique()

array([32])

 # remove space

In [ ]:

Order_Reviews_New['review_id'] = Order_Reviews_New['review_id'].str.strip()

In [ ]:
Order_Reviews_New['order_id'] = Order_Reviews_New['order_id'].str.strip()

In [ ]:
Order_Reviews_New.duplicated().sum()

np.int64(0)

# #checking duplicates in order_id and review_id

In [57]:
Order_Reviews_New['review_id'].duplicated().sum()

np.int64(814)

In [ ]:
Order_Reviews_New['order_id'].duplicated().sum()

np.int64(551)

In [ ]:
Order_Reviews_New['review_id'].nunique()

98410

# # remove duplicates in review_id

In [ ]:
Order_Reviews_New = Order_Reviews_New.drop_duplicates(subset=['review_id'])

In [ ]:
Order_Reviews_New.shape

(98410, 6)

In [ ]:
Order_Reviews_New['order_id'].isin(Order_Reviews_New['order_id']).sum()

np.int64(98410)

In [ ]:
Order_Reviews_New.duplicated().sum()

np.int64(0)

In [ ]:
Order_Reviews_New.isnull().sum()

,0
review_id,0
order_id,0
review_score,0
review_creation_date,0
review_answer_timestamp,0
response_time_hours,0


In [ ]:
Order_Reviews_New.to_csv("order_reviews_cleaned.csv", index=False)

In [ ]:
from google.colab import files
files.download("order_reviews_cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
Orders_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 10 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
 8   delivery_delay_days            96476 non-null  float64       
 9   delivery_days                  96476 non-null  float64       
dtypes: datetime64[ns](5), float64(2), object(3)
memory usage: 7.6+ MB


In [ ]:
Orders_New.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,delivery_delay_days
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:00,2017-10-02 11:07:00,2017-10-04 19:55:00,2017-10-10 21:25:00,2017-10-18,-8.0
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:00,2018-07-26 03:24:00,2018-07-26 14:31:00,2018-08-07 15:27:00,2018-08-13,-6.0
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:00,2018-08-08 08:55:00,2018-08-08 13:50:00,2018-08-17 18:06:00,2018-09-04,-18.0
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:00,2017-11-18 19:45:00,2017-11-22 13:39:00,2017-12-02 00:28:00,2017-12-15,-13.0
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:00,2018-02-13 22:20:00,2018-02-14 19:46:00,2018-02-16 18:17:00,2018-02-26,-10.0


In [ ]:
Orders_New['order_purchase_timestamp'] = pd.to_datetime(Orders_New['order_purchase_timestamp'])
Orders_New['order_approved_at'] = pd.to_datetime(Orders_New['order_approved_at'])
Orders_New['order_delivered_carrier_date'] = pd.to_datetime(Orders_New['order_delivered_carrier_date'])
Orders_New['order_delivered_customer_date'] = pd.to_datetime(Orders_New['order_delivered_customer_date'])
Orders_New['order_estimated_delivery_date'] = pd.to_datetime(Orders_New['order_estimated_delivery_date'])

In [ ]:
Orders_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 10 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
 8   delivery_delay_days            96476 non-null  float64       
 9   delivery_days                  96476 non-null  float64       
dtypes: datetime64[ns](5), float64(2), object(3)
memory usage: 7.6+ MB


In [ ]:
Orders_New['order_id'].duplicated().sum()

np.int64(0)

In [ ]:
Orders_New['customer_id'].duplicated().sum()

np.int64(0)

 # Most orders are:

delivered = 96,478

So about 97% of orders are completed.

In [ ]:
Orders_New['order_status'].value_counts()

,count
order_status,
delivered,96478
shipped,1107
canceled,625
unavailable,609
invoiced,314
processing,301
created,5
approved,2


In [ ]:
delivered_orders = Orders_New[Orders_New['order_status'] == 'delivered']

In [ ]:
Orders_New['delivery_delay_days'] = ( Orders_New['order_delivered_customer_date'] - Orders_New['order_estimated_delivery_date']).dt.days

In [ ]:
Orders_New['order_id'].str.len().unique()

array([32])

In [ ]:
Orders_New['customer_id'].str.len().unique()

array([32])

In [ ]:
Orders_New['order_id'] = Orders_New['order_id'].str.strip()

In [ ]:
Orders_New['customer_id'] = Orders_New['customer_id'].str.strip()

In [ ]:
Orders_New['order_id'].isnull().sum()

np.int64(0)

In [ ]:
Orders_New['order_estimated_delivery_date'].isnull().sum()

np.int64(0)

In [ ]:
Orders_New[Orders_New['order_delivered_customer_date'].isnull()]['order_status'].value_counts()

,count
order_status,
shipped,1107
canceled,619
unavailable,609
invoiced,314
processing,301
delivered,8
created,5
approved,2


In [ ]:
Orders_New[Orders_New['order_approved_at'].isnull()]['order_status'].value_counts()

,count
order_status,
canceled,141
delivered,14
created,5


In [ ]:
delivered_orders['delivery_delay_days'].mean()

np.float64(-11.875888877371203)

In [ ]:
Orders_New['delivery_days'] = (
    Orders_New['order_delivered_customer_date'] -
    Orders_New['order_purchase_timestamp']
).dt.days

 ## ##Average delivery time is approximately 12.09 days for delivered orders.
 ## ##Most orders are delivered within about 12 days after purchase, indicating the typical logistics time for order fulfillment.

In [ ]:
delivered_orders['delivery_days'].mean()

np.float64(12.09402923188556)

In [ ]:
Orders_New.shape

(99441, 10)

## ## The average delivery delay is approximately -11.88 days.
This indicates that orders are typically delivered about 12 days earlier than the estimated delivery date.

In [ ]:
delivered_orders['delivery_delay_days'].mean()

np.float64(-11.875888877371203)

In [ ]:
Orders_New.to_csv("Orders_cleaned.csv",index = False)

In [ ]:
from google.colab import files
files.download("Orders_cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [61]:
Customers_New.head(50)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


In [82]:
Customers_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [83]:
Customers_New.isnull().sum()

,0
customer_id,0
customer_unique_id,0
customer_zip_code_prefix,0
customer_city,0
customer_state,0


In [64]:
Customers_New.duplicated().sum()

np.int64(0)

In [65]:
Customers_New['customer_unique_id'].str.len().unique()

array([32])

In [76]:
Customers_New['customer_id'] = Customers_New['customer_id'].str.strip()

In [81]:
Customers_New['customer_state'].duplicated().sum()

np.int64(99414)

In [70]:
Customers_New = Customers_New.drop_duplicates(subset = ['customer_id'])

In [71]:
Customers_New['customer_city'] = Customers_New['customer_city'].str.strip().str.lower()
Customers_New['customer_state'] = Customers_New['customer_state'].str.strip().str.upper()

In [72]:
customers_orders = Customers_New.groupby('customer_unique_id')['customer_id'].count().reset_index()
customers_orders.rename(columns={'customer_id': 'total_orders'},inplace=True)

In [73]:
customers_orders.head()

,customer_unique_id,total_orders
0,0000366f3b9a7992bf8c76cfdf3221e2,1
1,0000b849f77a49e4a4ce2b2a4ca5be3f,1
2,0000f46a3911fa3c0805444483337064,1
3,0000f6ccb0745a6a4b88665a16c9f078,1
4,0004aac84e0df4da2b147fca70cf8255,1


In [74]:
customers_orders['customer_type'] = customers_orders['total_orders'].apply(lambda x: 'repeat_customer' if x > 1 else 'single_order_customer')

## ## Only about 3.12% of customers placed more than one order,
indicating low customer retention.

In [75]:
customers_orders['customer_type'].value_counts(normalize=True) * 100

,proportion
customer_type,
single_order_customer,96.881244
repeat_customer,3.118756


In [84]:
Customers_New.to_csv("Customers.csv",index=False)

In [85]:
from google.colab import files
files.download("Customers.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [113]:
Product_New.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_volume_cm3
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0,2240.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0,10800.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0,2430.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0,2704.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0,4420.0


In [132]:
Product_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32951 non-null  object 
 2   product_name_lenght         32951 non-null  int64  
 3   product_description_lenght  32951 non-null  int64  
 4   product_photos_qty          32951 non-null  int64  
 5   product_weight_g            32951 non-null  float64
 6   product_length_cm           32951 non-null  float64
 7   product_height_cm           32951 non-null  float64
 8   product_width_cm            32951 non-null  float64
 9   product_volume_cm3          32951 non-null  float64
dtypes: float64(5), int64(3), object(2)
memory usage: 2.5+ MB


In [112]:
Product_New.isnull().sum()

,0
product_id,0
product_category_name,0
product_name_lenght,0
product_description_lenght,0
product_photos_qty,0
product_weight_g,0
product_length_cm,0
product_height_cm,0
product_width_cm,0
product_volume_cm3,0


In [119]:
Product_New['product_id'].duplicated().sum()

np.int64(0)

In [89]:
Product_New.shape

(32951, 9)

In [90]:
Product_New['product_category_name'] = Product_New['product_category_name'].fillna('unknown')

In [92]:
Product_New['product_name_lenght'] = Product_New['product_name_lenght'].fillna(0)

In [93]:
Product_New['product_description_lenght'] = Product_New['product_description_lenght'].fillna(0)

In [94]:
Product_New['product_photos_qty'] = Product_New['product_photos_qty'].fillna(0)

In [96]:
Product_New['product_weight_g'] = Product_New['product_weight_g'].fillna(Product_New['product_weight_g'].median())

In [98]:
Product_New['product_length_cm'] = Product_New['product_length_cm'].fillna(Product_New['product_length_cm'].median())

In [100]:
Product_New['product_height_cm'] = Product_New['product_height_cm'].fillna(Product_New['product_height_cm'].median())

In [102]:
Product_New['product_width_cm'] = Product_New['product_width_cm'].fillna(Product_New['product_width_cm'].median())

In [110]:
Product_New['product_volume_cm3'] = (
    Product_New['product_length_cm'] *
    Product_New['product_height_cm'] *
    Product_New['product_width_cm']
)



In [125]:
Product_New['product_name_lenght'] = Product_New['product_name_lenght'].astype(int)
Product_New['product_description_lenght'] = Product_New['product_description_lenght'].astype(int)
Product_New['product_photos_qty'] = Product_New['product_photos_qty'].astype(int)

In [130]:
Product_New.to_csv("Product.csv",index=False)

In [131]:
from google.colab import files
files.download("Product.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [34]:
Order_Item_New.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,total_item_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:00,58.90,13.29,72.19
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:00,239.90,19.93,259.83
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:00,199.00,17.87,216.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:00,12.99,12.79,25.78
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:00,199.90,18.14,218.04


In [45]:
Order_Item_New.isnull().sum()

,0
order_id,0
order_item_id,0
product_id,0
seller_id,0
shipping_limit_date,0
price,0
freight_value,0
total_item_value,0


In [44]:
Order_Item_New['seller_id'].duplicated().sum()

np.int64(22376)

In [21]:
Order_Item_New.duplicated(subset=['order_id','order_item_id']).sum()

np.int64(0)

In [46]:
Order_Item_New.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24506 entries, 0 to 24505
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   order_id             24506 non-null  object        
 1   order_item_id        24506 non-null  int64         
 2   product_id           24506 non-null  object        
 3   seller_id            24506 non-null  object        
 4   shipping_limit_date  24506 non-null  datetime64[ns]
 5   price                24506 non-null  float64       
 6   freight_value        24506 non-null  float64       
 7   total_item_value     24506 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(3)
memory usage: 1.7+ MB


In [7]:
Order_Item_New[Order_Item_New.isnull().any(axis=1)]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
24506,37d6c8f1b209eab8d1b6522a1a3e4d88,1,111d0b1e69f3496156,NaN,NaN,NaN,NaN


In [8]:
Order_Item_New = Order_Item_New.dropna()

In [24]:
Order_Item_New ['shipping_limit_date'] = pd.to_datetime(Order_Item_New ['shipping_limit_date'])

## ## create total_item_value column

In [26]:
Order_Item_New['total_item_value'] = (Order_Item_New['price'] + Order_Item_New['freight_value'])

In [29]:
Order_Item_New.shape

(24506, 8)

In [36]:
Order_Item_New['seller_id'].str.len().unique()

array([32])

In [39]:
Order_Item_New['seller_id'] = Order_Item_New['seller_id'].str.strip()

In [47]:
Order_Item_New.to_csv("Order_Item.csv",index=False)

In [48]:
from google.colab import files
files.download("Order_Item.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [67]:
Order_Payments_New.head()


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [77]:
Order_Payments_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [71]:
Order_Payments_New.isnull().sum()

,0
order_id,0
payment_sequential,0
payment_type,0
payment_installments,0
payment_value,0


In [73]:
Order_Payments_New['order_id'] = Order_Payments_New['order_id'].str.strip()

In [74]:
Order_Payments_New['order_id'].str.len().unique()

array([32])

In [75]:
Order_Payments_New.duplicated().sum()

np.int64(0)

In [79]:
Order_Payments_New['order_id'].duplicated().sum()

np.int64(4446)

In [57]:
Order_Payments_New['payment_type'].value_counts()

,count
payment_type,
credit_card,76795
boleto,19784
voucher,5775
debit_card,1529
unknown,3


In [56]:
Order_Payments_New['payment_type'] = Order_Payments_New['payment_type'].replace('not_defined','unknown')

In [78]:
Order_Payments_New.duplicated(subset=['order_id','payment_sequential']).sum()

np.int64(0)

In [82]:
Order_Payments_New.to_csv("Orders_Payments.csv",index=False)

In [83]:
from google.colab import files
files.download("Orders_Payments.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [85]:
Sellers_New.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [100]:
Sellers_New.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [95]:
Sellers_New['seller_id'].str.len().unique()

array([32])

In [96]:
Sellers_New['seller_id'] = Sellers_New['seller_id'].str.strip()

In [97]:
Sellers_New.isnull().sum()

,0
seller_id,0
seller_zip_code_prefix,0
seller_city,0
seller_state,0


In [88]:
Sellers_New.duplicated().sum()

np.int64(0)

In [92]:
Sellers_New['seller_city'] = Sellers_New['seller_city'].str.strip().str.lower()



In [93]:
Sellers_New['seller_state'] = Sellers_New['seller_state'].str.strip().str.upper()

In [94]:
Sellers_New['seller_state'].value_counts()

,count
seller_state,
SP,1849
PR,349
MG,244
SC,190
RJ,171
RS,129
GO,40
DF,30
ES,23


In [99]:
Sellers_New['seller_id'].duplicated().sum()

np.int64(0)

In [103]:
Sellers_New.to_csv("Sellers.csv",index=False)

In [104]:
from google.colab import files
files.download("Sellers.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [106]:
GeoLocation_New.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [126]:
GeoLocation_New.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19015 entries, 0 to 999846
Data columns (total 5 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   geolocation_zip_code_prefix  19015 non-null  int64  
 1   geolocation_lat              19015 non-null  float64
 2   geolocation_lng              19015 non-null  float64
 3   geolocation_city             19015 non-null  object 
 4   geolocation_state            19015 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 891.3+ KB


In [125]:
GeoLocation_New.isnull().sum()

,0
geolocation_zip_code_prefix,0
geolocation_lat,0
geolocation_lng,0
geolocation_city,0
geolocation_state,0


In [109]:
GeoLocation_New['geolocation_city'] = GeoLocation_New['geolocation_city'].str.strip().str.lower()

In [110]:
GeoLocation_New['geolocation_state'] = GeoLocation_New['geolocation_state'].str.strip().str.upper()

In [124]:
GeoLocation_New.duplicated().sum()

np.int64(0)

In [123]:
GeoLocation_New['geolocation_state'].duplicated().sum()

np.int64(18988)

In [113]:
GeoLocation_New = GeoLocation_New.drop_duplicates(subset = ['geolocation_zip_code_prefix'])

In [127]:
GeoLocation_New.to_csv("GeoLocation.csv",index=False)

In [128]:
from google.colab import files
files.download("GeoLocation.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>